### ISSUE TITLE
OS X build steps

### ISSUE THREAD

**Issue by phries** [jcelaya/hdrmerge#56]
Here's what I've found for running HDRMerge on Mac OS X 10.10 (Yosemite). Not totally successful yet but I got it built and running.
### Build

Uses [brew package manager](http://brew.sh/). I had a few brews already installed that might have been required too.
- brew install libraw
- brew install cmake
- brew install qt4
- brew install boost
- Download and build the tarball of [exiv2](http://www.exiv2.org/archive.html) (0.23, not latest). Install it with: sudo make install
##### alglib (long step for _develop branch only_)

For the "develop" branch, you also need alglib 3.x.x. The only homebrew formula for "alglib" gives you alglib 2 which is not compatible. Complicating things, neither version of alglib includes any kind of Makefile, so every package manager rolls their own.

For now you can use a [quick homebrew formula](https://gist.github.com/phries/964ac206335a251c934f) I wrote for "alglib3":
- brew install https://gist.githubusercontent.com/phries/964ac206335a251c934f/raw/3cbac1af5dd2ecced8f6d46f29e050074db62f31/alglib3.rb

And apply this diff to hdrmerge:

```
$ git diff
diff --git a/cmake/FindAlglib.cmake b/cmake/FindAlglib.cmake
index 9b80b9a..5903414 100644
--- a/cmake/FindAlglib.cmake
+++ b/cmake/FindAlglib.cmake
@@ -26,10 +26,10 @@ find_path (ALGLIB_INCLUDES
     statistics.h
     stdafx.h
     PATHS
-    /usr/include/libalglib/
+    /usr/local/include/alglib3/
     )   

-find_library (ALGLIB_LIBRARIES NAMES alglib)
+find_library (ALGLIB_LIBRARIES NAMES alglib3)

 # handle the QUIETLY and REQUIRED arguments and set ALGLIB_FOUND to TRUE if
```
##### Compiling hdrmerge

In build dir:
- cmake ..
- make
### Issues
- GUI app crashes in hdrmerge's alignment code when I try to merge the photos.
- Command line creates a merged photo but it's black (see #52). This may not be Mac-related.

---

_edit 11/30 & 12/5: Worked around earlier need for installing a new compiler, which is time consuming. The program is much more debuggable with the default (XCode) compiler anyway. Compile from my fork, "develop" branch for now if you have compilation errors in things like an algorithm include and lambdas_

**Comment by jcelaya** [comment]
Thank you for your effort, Phries :)
I managed to build hdrmerge without problem, with latest XCode and brew. I will test it and try to build an installation package to include it with the next stable release. I may also switch Alglib with GSL, which is included in brew.

**Comment by Beep6581** [comment]
@phries @benitoite we intend to release a new version soon, and it would be great if you could provide a dev build for testing, and instructions how to compile it in recent macOS versions.

**Comment by Benitoite** [comment]
@Beep6581 https://filebin.net/6q2gbc5wt6jkv52h/HDRMerge-test1.dmg

These steps don't include things like getting Xcode, a MacOS SDK, Qt4, or cmake, etc. Don't have homebrew, nor do I use Xcode extensively, so I did the following to build HDRMerge on the command line:

1a. Dependencies. Get the dependencies which give you these libraries. Some are from MacPorts and some are built from their code repositories.  
```
libexiv2.26.dylib
libexpat.1.dylib
libiconv.2.dylib
libintl.8.dylib
libiomp5.dylib
libjasper.4.dylib
libjpeg.9.dylib
liblcms2.2.dylib
libmng.2.dylib
libpng16.16.dylib
libraw_r.16.dylib
libtiff.5.dylib
libz.1.dylib
```
1b. HDRMerge is cloned into ~/hdrmerge, ALGLIB 3.13.0 for C++ is [downloaded](http://www.alglib.net/download.php) and extracted to ~/alglib

2. Copy alglib files to hdrmerge:
```
cp ~/alglib/cpp/src/* ~/hdrmerge/.
```
3. Update CMakeLists.txt to allow non-Apple clang, include alglib files, and install to ~/hdrmerge/build/release instead of /Applications
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index b3db8f0..18f6b4e 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,5 @@
+CMAKE_POLICY(SET CMP0025 NEW)
 cmake_minimum_required(VERSION 2.8)
-
 project(hdrmerge)
 
 # Complain about anything, except ignored OpenMP pragma in Debug
@@ -74,7 +74,8 @@ endif (WIN32)
 # Sources and headers
 set(hdrmerge_sources
     Image.cpp ImageStack.cpp Bitmap.cpp RawParameters.cpp EditableMask.cpp DngFloatWriter.cpp
-    TiffDirectory.cpp BoxBlur.cpp ExifTransfer.cpp ImageIO.cpp)
+    TiffDirectory.cpp BoxBlur.cpp ExifTransfer.cpp ImageIO.cpp ap.cpp integration.cpp interpolation.cpp alglibinternal.cpp linalg.cpp
+    alglibmisc.cpp specialfunctions.cpp solvers.cpp optimization.cpp)
 set(hdrmerge_gui_sources
     AboutDialog.cpp MainWindow.cpp PreviewWidget.cpp DraggableScrollArea.cpp DngPropertiesDialog.cpp LoadOptionsDialog.cpp)
 set(hdrmerge_qobject_headers MainWindow.hpp PreviewWidget.hpp LoadOptionsDialog.hpp
@@ -135,7 +136,7 @@ if (WIN32)
     endif (MAKENSIS_EXECUTABLE)
 elseif (APPLE)
     set_source_files_properties(images/icon.icns PROPERTIES MACOSX_PACKAGE_LOCATION "Resources")
-    install(TARGETS hdrmerge BUNDLE DESTINATION "/Applications")
+    install(TARGETS hdrmerge BUNDLE DESTINATION "~/hdrmerge/build/release")
 else (WIN32)
     install(TARGETS hdrmerge RUNTIME DESTINATION bin)
     find_program(XDG-DESKTOP-MENU_EXECUTABLE xdg-desktop-menu)
```
4a. Make and goto the build directory. 
```
mkdir ~/hdrmerge/build && cd ~/hdrmerge/build
```
4b. Issue cmake command
```
sudo cmake .. -DQT_QMAKE_EXECUTABLE=/opt/local/libexec/qt4/bin/qmake -DCMAKE_BUILD_TYPE=Release -DOpenMP_C_FLAGS=-fopenmp="libiomp5" -DOpenMP_CXX_FLAGS=-fopenmp="libiomp5" -DOpenMP_C_LIB_NAMES="libiomp5" -DOpenMP_CXX_LIB_NAMES="libiomp5" -DOpenMP_libiomp5_LIBRARY="/opt/local" -DALGLIB_INCLUDES=~/hdrmerge -DALGLIB_LIBRARIES=~/hdrmerge -DCMAKE_C_COMPILER="clang-mp-3.9"       -DCMAKE_CXX_COMPILER="clang++-mp-3.9" -DCMAKE_OSX_SYSROOT="/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk"
```
5. compile
```
sudo make -j4 install
```
6. Create directory for Frameworks
```
sudo mkdir ~/hdrmerge/build/release/hdrmerge.app/Contents/Frameworks
```
7. Copy two of the dependencies into Frameworks.
```
sudo cp /opt/local/lib/libomp/libiomp5.dylib ~/hdrmerge/build/release/hdrmerge.app/Contents/Frameworks/.
sudo cp /usr/local/lib/libexiv2.26.dylib ~/hdrmerge/build/release/hdrmerge.app/Contents/Frameworks/.
```
8. Run Qt4's macdeployqt twice in a row.
```
sudo  /opt/local/libexec/qt4/bin/macdeployqt ~/hdrmerge/build/release/hdrmerge.app -no-strip -verbose=3
sudo  /opt/local/libexec/qt4/bin/macdeployqt ~/hdrmerge/build/release/hdrmerge.app -no-strip -verbose=3
```
9. Install an rpath
```
sudo install_name_tool -add_rpath "@executable_path/.." ~/hdrmerge/build/release/hdrmerge.app/Contents/MacOS/hdrmerge
```
10. Make the .dmg
```
sudo hdiutil create -ov -srcfolder ~/hdrmerge/build/release/hdrmerge.app ~/hdrmerge/build/release/HDRMerge.dmg
```

Some things are optional, like libiomp5 instead of clang's libomp. I don't think libiomp5 is required, but it is what I have used. After building other software I find I already had the dependencies except alglib, which took a few minutes to figure out how to incorporate. Also note I'm using the MacOS X 10.9 SDK. App test run done in 10.11 and 10.12, should run on 10.9+.

:barber:

**Comment by Beep6581** [comment]
@Benitoite great, I opened a forum for testing:
https://discuss.pixls.us/t/help-test-the-new-hdrmerge-development-build-for-macos/6936

@Benitoite did you compile using the `master` or the `treeclean` branch?

**Comment by Benitoite** [comment]
I built the master branch but can try with the other one later today.

**Comment by Beep6581** [comment]
@Benitoite I will merge `treeclean` into `master` soon, @Hombre57 is still going to make a commit before I do, so maybe better if you wait for that. I will ping you when it's done.

**Comment by aferrero2707** [comment]
I have downloaded the .dmg file, and will start testing it soon...

On a somewhat related note, I have some experience to offer in automating the OSX building/packaging using Travis CI. I have such an automated procedure already set-up and running for [another project](https://github.com/aferrero2707/PhotoFlow/blob/stable/.travis.yml#L192). It is based on homebrew, and the builds can be either triggered by commits to the relevant branches, or started regularly with CRON jobs...

**Comment by Beep6581** [comment]
@Benitoite @aferrero2707 branch `treeclean` merged into `master`. You can go ahead and test macOS compilation. Feel free to make the required changes to the CMake file, but note that @Hombre57 is also making changes to the Windows sections of that file.

**Comment by Benitoite** [comment]
@Beep6581 MacOS build commands and CMakeLists.txt changes above updated.

**Comment by Beep6581** [comment]
@Benitoite to prepare for release v1.0, are the macOS build instructions still valid?

**Comment by Benitoite** [comment]
@Beep6581 https://github.com/jcelaya/hdrmerge/pull/152 Just had to update CMakeLists.txt to point to the icon in data/images.  Updated instructions above for Qt 5.11.0.

**Comment by dannephoto** [comment]
Hi. I tried compiling form time to time but have not been successful until now. Following phries but mostly Benitoite I used the steps to build an automator application dedicated for compilation on mac.
https://bitbucket.org/Dannephoto/hdrmerge_compiler
![screen shot 2018-06-25 at 15 58 19 png_500px](https://user-images.githubusercontent.com/22038358/41856684-fda4c4f0-7895-11e8-9dbf-aeffc0098b80.jpg)

The compiler will check for dependencies and install what is missing, then present the menu above. Script source:
https://bitbucket.org/Dannephoto/hdrmerge_compiler/src/default/HDRMerge_compiler.app/Contents/main.command

I changed alglib to version 14 and the exiv2 files are fetched from a brew install. QT version used is also coming from brew. Tried exchanging macports clang-mp-3.9 dependency to brew llvm@5 and it works kind of but there will be some libomp issues/questions so reverted back to macport again.

Ok, A couple of questions about the cmake command. I removed following from the command but it still seems to work here. not needed?
```
#not needed?
#-DQt5_DIR=$HOME/Qt/5.11.0/clang_64/lib/cmake/Qt5 
#-DCMAKE_OSX_SYSROOT="/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk"
```

My cmake command goes like this:
`sudo cmake .. -DCMAKE_BUILD_TYPE=Release -DOpenMP_C_FLAGS=-fopenmp="libiomp5" -DOpenMP_CXX_FLAGS=-fopenmp="libiomp5" -DOpenMP_C_LIB_NAMES="libiomp5" -DOpenMP_CXX_LIB_NAMES="libiomp5" -DOpenMP_libiomp5_LIBRARY="/opt/local" -DCMAKE_C_COMPILER="clang-mp-3.9" -DCMAKE_CXX_COMPILER="clang++-mp-3.9" -DALGLIB_ROOT=$ALGLIB_ROOT -DALGLIB_INCLUDES=$ALGLIB_ROOT/src -DALGLIB_LIBRARIES=$ALGLIB_ROOT/src -DCMAKE_INSTALL_BINDIR=$HOME/hdrmerge/build/install`

Thanks again for an amazing application.

**Comment by Benitoite** [comment]
@dannephoto Don't hesitate to ask if you have any other issues compiling.
 The warning regarding alglib wanting libraries is ignorable, as binaries are compiled during the `make`.  
The `not needed?` commands might indeed be redundant for brew users, but from an installation of qt from the qt.io installer cmake seems to need the pointing there. 
Also, if compiling for self use, you can leave out the SYSROOT assignment. The SYSROOT is a part of Xcode that let's you be compatible with previous macOS versions, in the given example 10.9.

**Comment by dannephoto** [comment]
Thanks. Mainly for my own amusement :).
Quite some learning experience getting this going actually but important since it's such a nice tool.

**Comment by dannephoto** [comment]
@Benitoite 
One issue which made me exclude the SYSROOT part was this:
```
  ld: unexpected token: !tapi-tbd-v2 file
  '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/lib/libSystem.tbd'
  for architecture x86_64

  clang: error: linker command failed with exit code 1 (use -v to see
  invocation)

  make[1]: *** [cmTC_f5f31] Error 1

  make: *** [cmTC_f5f31/fast] Error 2
```
Related to version being MacOSX10.12.sdk?

**Comment by Benitoite** [comment]
Yes I’ve seen this issue with Xcode 9.2’s ‘ld’ - someone was trying to compile rawtherapee and got the same ‘!tapi’ error.  They figured it out to the exact command in your solution and thought that ought to be in the compilation instructions; will have to incorporate the command in HDRMerge instructions.

**Comment by dannephoto** [comment]
Tested on a vanilla mac(missing pkgconfig). For completeness:
#pkgconfig
if ! [ -d /usr/local/Cellar/pkg-config ]; then
brew install pkg-config
fi

**Comment by fanckush** [comment]
Sorry for reviving an old issue! I though I'd add potentially useful info to this.

I followed the steps from https://github.com/jcelaya/hdrmerge/issues/56#issuecomment-372105553 but changed the Cmake commend a bit
Here's the cmake command I used on macOS **Mojave** 10.14.4 and using **brew**
```zsh
cmake .. -DQt5_DIR=/usr/local/Cellar/Qt/5.12.3/lib/cmake/Qt5 -DCMAKE_BUILD_TYPE=Release -DOpenMP_C_FLAGS=-fopenmp=lomp -DOpenMP_CXX_FLAGS=-fopenmp=lomp -DOpenMP_C_LIB_NAMES="libiomp5" -DOpenMP_CXX_LIB_NAMES="libiomp5" -DOpenMP_libiomp5_LIBRARY="/usr/local/lib/libomp.dylib" -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp /usr/local/lib/libomp.dylib -I/usr/local/include" -DOpenMP_CXX_LIB_NAMES="libiomp5" -DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp /usr/local/lib/libomp.dylib -I/usr/local/include"   -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_C_COMPILER="clang" -DCMAKE_CXX_COMPILER="clang++" -DCMAKE_OSX_SYSROOT="/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk" -DALGLIB_ROOT=$ALGLIB_ROOT -DALGLIB_INCLUDES=$ALGLIB_ROOT/src -DALGLIB_LIBRARIES=$ALGLIB_ROOT/src -DCMAKE_INSTALL_BINDIR=$HOME/hdrmerge/build/install  
```

**Comment by Benitoite** [comment]
` -DOpenMP_CXX_LIB_NAMES="libiomp5"`  a _libiomp5_ still in there.  Is there a brew command to install everything or partially what's needed?

**Comment by Beep6581** [comment]
Would be nice if someone updated the macOS section in  https://github.com/jcelaya/hdrmerge/blob/master/INSTALL.md as needed.

**Comment by dannephoto** [comment]
@fanckush 
How are you getting '/usr/local/lib/libomp.dylib' installed?

I get this with your set up.
```
clang: error: no such file or directory: '/usr/local/lib/libomp.dylib'
clang: error: no such file or directory: '/usr/local/lib/libomp.dylib'
clang: error: no such file or directory: '/usr/local/lib/libomp.dylib'

Running my hdr_merge compiler I run following and working:
`sudo cmake .. -DCMAKE_BUILD_TYPE=Release -DOpenMP_C_FLAGS=-fopenmp="libiomp5" -DOpenMP_CXX_FLAGS=-fopenmp="libiomp5" -DOpenMP_C_LIB_NAMES="libiomp5" -DOpenMP_CXX_LIB_NAMES="libiomp5" -DOpenMP_libiomp5_LIBRARY="/opt/local" -DCMAKE_C_COMPILER="clang-mp-3.9" -DCMAKE_CXX_COMPILER="clang++-mp-3.9" -DALGLIB_ROOT=$ALGLIB_ROOT -DALGLIB_INCLUDES=$ALGLIB_ROOT/src -DALGLIB_LIBRARIES=$ALGLIB_ROOT/src -DCMAKE_OSX_SYSROOT="$SDK" -DCMAKE_INSTALL_BINDIR=$HOME/hdrmerge/build/install`
Source: https://bitbucket.org/Dannephoto/hdrmerge_compiler/src/default/HDRMerge_compiler.app/Contents/main.command
```

**Comment by fanckush** [comment]
@Beep6581 i’ll work on that for brew users

**Comment by fanckush** [comment]
Can someone test the new instructions from: #179 
I tested on two machines:
- Mojave 10.14.4
- High Sierra 10.13.x (without Xcode)

**Comment by heckflosse** [comment]
@fanckush I can not test as I'm not an Mac, but I placed two typo comments on your pr.

**Comment by fanckush** [comment]
@Beep6581 I think this can be closed now that #179 is merged
